src/coreclr/jit/emitxarch.cpp

```diff
@@ -14691,13 +14708,13 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
 
     assert((int)emitCurStackLvl >= 0);
 
-    // Only epilog "instructions" and some pseudo-instrs
+    // Only epilog "instructions", some pseudo-instrs and block ending jumps
     // are allowed not to generate any code
 
-    assert(*dp != dst || emitInstHasNoCode(ins));
+    assert(*dp != dst || emitInstHasNoCode(id));
 
 #ifdef DEBUG
-    if (emitComp->opts.disAsm || emitComp->verbose)
+    if ((emitComp->opts.disAsm || emitComp->verbose) && !emitInstHasNoCode(id))
```

Comment:
After some debugging of your branch, found out that the addition of `!emitInstHasNoCode(id)` is the reason why we stopped displaying `align` instruction in the disassembly and that is the reason we were seeing the `3 bytes` difference.

![image](https://user-images.githubusercontent.com/12488060/171986790-d969d519-9d0f-439f-a2e8-2e6e6a8b5639.png)

You will have to separate out the conditions for `jmp` and `align` and use the combination only at common places but use the `jmp` condition in jump specific code-paths.

Here is something that you might want to consider: https://github.com/kunalspathak/runtime/commit/aaa8f70b33f607e3960173246bbd078f65358075

Reply:
Ok. pushed.

If the group isn't marked as align and there zero alignment bytes to be allocated what is the purpose of printing the instruction?

Reply:
Retaining or removing of `align` instructions are heuristic sensitive vs. `jmp` instructions removal which is correctness sensitive. During perf investigation, it is often useful to see a sensitive loop that was marked for alignment but was later decided to not do it by spotting those `align [0 bytes]` in disassembly. It turns out useful to spot such cases and adjust the heuristics to make them align. We also add `NOP` compensation (if needed) until the last IG that needs alignment.  By printing the zero size `align`, it help us to reason out the purpose of compensation instructions added.